### PR TITLE
Added decorators for custom combiners, losses, metrics

### LINKED ISSUE #1502: Extensible loss and metric modules
Currently, users need to make changes to the underlying Ludwig internals in order to add a new loss or metric. Similar to what we did for encoders / decoders, it should be possible for users to add and register new losses and metrics using a simple decorator system like the following:

```
@loss("my_loss")
class MyLoss(nn.Module):
    def forward(self, preds: Tensor, target: Tensor) -> Tensor:
        ...

@metric("my_metric")
class MyMetric(Metric):
    def update(self, preds: Tensor, target: Tensor) -> None:
        ...

    def compute(self) -> Tensor:
        ...

    @abstractmethod
    def get_current_value(self, preds: Tensor, target: Tensor) -> Tensor:
        ...

@register_loss("my_loss", [NUMERICAL, CATEGORY])
@register_metric("my_metric", [NUMERICAL])
```

### PR DESCRIPTION
**Pull request by tgaddair**
Any class can now be made a custom combiner, loss, or metric as follows:

```
@register("my_combiner")
class MyCombiner:
    ...

@register_loss("my_loss", [NUMERICAL, VECTOR])
class MyLoss:
    ...

@register_metric("my_metric", [CATEGORY])
class MyMetric:
    ...
```

Also refactors Encoders to use a new `register_encoder` decorator. This allows us to remove the special subclasses of `Encoder` that only serve to put the derived classes into the correct registry. Same for Decoders.

Fixes #1502.

### CODE REVIEW

File: ludwig/combiners/combiners.py [hunk 1]
```diff
@@ -482,19 +491,18 @@ class Meta:
         unknown = INCLUDE
 
 
-class TabNetCombiner(CombinerClass):
+@register_combiner(name='tabnet')
+class TabNetCombiner(Combiner):
     def __init__(
             self,
-            input_features: Dict,
+            input_features: Dict[str, "InputFeature"],
```
Comment (justinxzhao) [inline ludwig/combiners/combiners.py:498]: Is wrapping the class name in quotes the standard way to declare a type hint for a Dict input arg?

File: ludwig/combiners/combiners.py [hunk 2]
```diff
@@ -1131,14 +1122,3 @@ def get_combiner_class(combiner_type):
         combiner_type,
         combiner_registry
     )
-
-
-combiner_registry = {
```
Comment (justinxzhao) [inline ludwig/combiners/combiners.py:1136]: Nice!

**Review by justinxzhao — APPROVED** [review]
LGTM! Feels like a big improvement over the current registry mechanisms.

File: ludwig/combiners/combiners.py (hunk 1 above)
Comment (tgaddair) [reply ludwig/combiners/combiners.py:498]: It's a way to reference the type without needing to import it (in this case, it would be a circular dependency).
